server/src/main/java/org/apache/iotdb/db/query/executor/LastQueryExecutor.java

```diff
@@ -0,0 +1,216 @@
+    dataSet.setRecord(resultRecord);
+    return dataSet;
+  }
+
+  /**
+   * get aggregation result for one series
+   *
+   * @param context query context
+   * @return AggregateResult list
+   */
+  private LastQueryResult calculateLastPairForOneSeries(
+          Path seriesPath, TSDataType tsDataType,
+          QueryContext context)
+          throws IOException, QueryProcessException, StorageEngineException {
+    LastQueryResult queryResult = new LastQueryResult();
+    MNode node = null;
+    try {
+      node = MManager.getInstance().getNodeByPathFromCache(seriesPath.toString());
+    } catch (PathException e) {
+      throw new QueryProcessException(e);
+    } catch (CacheException e) {
+      throw new QueryProcessException(e.getMessage());
+    }
+    if (node.getCachedLast() != null) {
+      queryResult.setPairResult(node.getCachedLast());
+      return queryResult;
+    }
+
+    // construct series reader without value filter
+    Filter timeFilter = null;
+    IAggregateReader seriesReader = new SeriesAggregateReader(
+            seriesPath, tsDataType, context, QueryResourceManager.getInstance()
+            .getQueryDataSource(seriesPath, context, timeFilter), timeFilter, null);
+
+    long maxTime = Long.MIN_VALUE;
+    while (seriesReader.hasNextChunk()) {
+      // cal by chunk statistics
+      if (seriesReader.canUseCurrentChunkStatistics()) {
+        Statistics chunkStatistics = seriesReader.currentChunkStatistics();
+        if (chunkStatistics.getEndTime() > maxTime) {
+          maxTime = chunkStatistics.getEndTime();
+          queryResult.setPairResult(maxTime, chunkStatistics.getLastValue(), tsDataType);
+        }
+        seriesReader.skipCurrentChunk();
+        continue;
+      }
+      while (seriesReader.hasNextPage()) {
+        //cal by page statistics
+        if (seriesReader.canUseCurrentPageStatistics()) {
+          Statistics pageStatistic = seriesReader.currentPageStatistics();
+          if (pageStatistic.getEndTime() > maxTime) {
+            maxTime = pageStatistic.getEndTime();
+            queryResult.setPairResult(maxTime, pageStatistic.getLastValue(), tsDataType);
+          }
+          seriesReader.skipCurrentPage();
+          continue;
+        }
+        // cal by page data
+        while (seriesReader.hasNextOverlappedPage()) {
+          BatchData nextOverlappedPageData = seriesReader.nextOverlappedPage();
+          int maxIndex = nextOverlappedPageData.length() - 1;
+          if (maxIndex < 0) {
+            continue;
+          }
+          long time = nextOverlappedPageData.getTimeByIndex(maxIndex);
+          if (time > maxTime) {
+            maxTime = time;
+            queryResult.setPairResult(maxTime, nextOverlappedPageData.getValueInTimestamp(time), tsDataType);
+          }
+          nextOverlappedPageData.resetBatchData();
+        }
+      }
+    }
+    if (queryResult.hasResult())
+      node.setCachedLast(queryResult.getPairResult());
```

Comment:
Notice:  considering inserting a point while querying. 

It's better to remove the setCachedLast and change updateCachedLast method to a synchronized one. Then, extend updateCachedLast with a parameter: boolean highPriority, which is to determine whether update the value if the time is equal to the previous cached one.

Then, in the normal inserting process, update the cache with high priority. Here, update the cache with low priority.
This could resolve the condition: there is an insert (10,11) when you get the last point (10,10) from disk.

Reply:
Updated, thanks